src/simulate-swap.ts

```diff
@@ -0,0 +1,927 @@
+    const liquidity = pool.liquidity
+    sqrtPrice = pool.sqrtPrice
+    const [swapLimit, hasLimitingTick, limitingTickIndex, isInitialized, tickLimitReached] =
+      getCloserLimit(sqrtPriceLimit, xToY, currentTickIndex, tickSpacing, tickmap)
+
+    if (tickLimitReached) {
+      globalInsufficientLiquidity = true
+      break
+    }
+
+    const swapResult = computeSwapStep(
+      sqrtPrice,
+      swapLimit,
+      liquidity,
+      remainingAmount,
+      byAmountIn,
+      feeTier.fee.v
+    )
+
+    if (byAmountIn) {
+      remainingAmount -= swapResult.amountIn + swapResult.feeAmount
+    } else {
+      remainingAmount -= swapResult.amountOut
+    }
+
+    // pool = await AddFee(pool, swapResult.feeAmount, xToY, protocolFee)
+    pool.sqrtPrice = swapResult.nextSqrtPrice
+
+    totalAmountIn += swapResult.amountIn + swapResult.feeAmount
+    totalAmountOut += swapResult.amountOut
+    eventFeeAmount += swapResult.feeAmount
+
+    // fail if price would go over swap limit
+    if (swapResult.nextSqrtPrice === sqrtPriceLimit && remainingAmount != 0n) {
+      globalInsufficientLiquidity = true
+      break
+    }
+
+    let tick: TickVariant | undefined
+    if (hasLimitingTick) {
+      if (isInitialized) {
+        tick = ticks.find(t => t.index === limitingTickIndex)
+        // tick not found despite of being reported as initialized - state mismatch
+        if (!tick) {
+          stateOutdated = true
+          break
+        }
+      }
+    }
+
+    const [amountToAdd, amountAfterTickUpdate, hasCrossed, overOrUnderflow] = poolUpdateTick(
+      pool,
+      tick,
+      swapResult.nextSqrtPrice,
+      swapLimit,
+      remainingAmount,
+      byAmountIn,
+      xToY,
+      hasLimitingTick,
+      isInitialized,
+      limitingTickIndex
+    )
+    stateOutdated ||= overOrUnderflow
+    if (stateOutdated) {
+      break
+    }
+
+    remainingAmount = amountAfterTickUpdate
+    totalAmountIn += amountToAdd
+
+    if (hasCrossed && tick) {
+      ticksCrossed.push(tick)
+      if (ticksCrossed.length > MAX_TICK_CROSS) {
+        maxTicksCrossed = true
+        break
+      }
+    }
+
+    const reachedTickLimit = xToY
+      ? pool.currentTickIndex <= tickLimit
+      : pool.currentTickIndex >= tickLimit
+
+    if (reachedTickLimit) {
+      globalInsufficientLiquidity = true
+      break
+    }
+  }
+
+  if (totalAmountOut === 0n) {
+    throw new Error(String(InvariantError.NoGainSwap))
+  }
+
+  return {
+    amountIn: totalAmountIn,
+    amountOut: totalAmountOut,
+    startSqrtPrice: eventStartSqrtPrice,
+    targetSqrtPrice: pool.sqrtPrice,
+    fee: eventFeeAmount,
+    crossedTicks: ticksCrossed,
+    globalInsufficientLiquidity,
+    stateOutdated,
+    maxTicksCrossed
+  }
+}
+
+const getCloserLimit = (
+  sqrtPriceLimit: bigint,
+  xToY: boolean,
+  currentTickIndex: bigint,
+  tickSpacing: bigint,
+  tickmap: Tickmap
+): [bigint, boolean, bigint, boolean, boolean] => {
```

Comment:
I agree that any tuple that has more than 3 elements should be replaced with some sort of structured object. This could also be considered the next time we have a PR tweaking the contract to apply that logic there (and we will have one after the next Alephium update due to the added scoping and a currently benign bug I found).